src/config/config.h

```diff
@@ -0,0 +1,243 @@
+    std::shared_ptr<Config<data_t>> Get(const std::string& config_name, data_t&& default_value);
+
+   private:
+    using config_base_ptr_t = std::shared_ptr<ConfigBase>;
+    using config_map_t      = std::unordered_map<std::string, config_base_ptr_t>;
+
+    struct JsonParsingContext {
+        simdjson::ondemand::parser   parser_;
+        simdjson::padded_string      buf_;
+        simdjson::ondemand::document doc_;
+    };
+
+    std::shared_ptr<ConfigBase> RegisterOrGet(const std::string& config_name, std::shared_ptr<ConfigBase>&& config);
+
+    void InitConfigJsonContext();
+
+    std::string        filepath_;
+    JsonParsingContext json_context_;
+    std::mutex         map_mutex_;
+    config_map_t       config_map_;
+};
+
+template<class data_t>
+Config<data_t>::Config(const std::string& name) : name_(name) {
+}
+
+namespace impl {
+
+template<class data_t>
+std::string ConfigDataGetStringRep(const data_t&) {
+    return core::GetTypeName<data_t>();
+}
+
+template<>
+inline std::string ConfigDataGetStringRep(const std::string& data) {
+    // Ignoring escaping for simplicity.
+    return "\"" + data + "\"";
+}
+
+template<class data_t>
+    requires requires(data_t data) { std::to_string(data); }
+std::string ConfigDataGetStringRep(const data_t& data) {
+    return std::to_string(data);
+}
+
+template<class data_t>
+    requires requires(data_t data) { std::string(data.Str()); }
+std::string ConfigDataGetStringRep(const data_t& data) {
+    return data.Str();
+}
+
+template<class data_t>
+    requires requires(data_t data) { std::string(data->Str()); }
+std::string ConfigDataGetStringRep(const data_t& data) {
+    return data->Str();
+}
+
+template<class T1, class T2>
+std::string ConfigDataGetStringRep(const std::pair<T1, T2>& data) {
+    return "[" + ConfigDataGetStringRep(data.first) + ", " + ConfigDataGetStringRep(data.second) + "]";
+}
+
+template<class data_t>
+    // clang-format off
+    requires(!std::same_as<data_t, std::string>) && requires(data_t data) {
+        std::accumulate(
+            data.begin(),
+            data.end(),
+            std::declval<std::string>(),
+            std::declval<std::function<std::string(std::string, const typename data_t::value_type&)>>());
+    }
+// clang-format on
+std::string ConfigDataGetStringRep(const data_t& data) {
+    return std::accumulate(
+               data.begin(),
+               data.end(),
+               std::string(),
+               [](std::string result, const auto& item) {
+                   result += result.empty() ? "[" : ",";
+                   return std::move(result) + ConfigDataGetStringRep(item);
+               }) +
+        "]";
+}
+
+}  // namespace impl
+
+template<class data_t>
+std::shared_ptr<Config<data_t>> ConfigFile::Get(const std::string& config_name) {
+    return Get(config_name, data_t{});
+}
+
+template<class data_t>
+std::shared_ptr<Config<data_t>> ConfigFile::Get(const std::string& config_name, data_t&& default_value) {
```

Comment:
Done by adding config type concept.

Reply:
Looks like you blocked deduction to T&?
But what's wrong with defaulting to lvref?

Reply:
We will move the value to `data_`, that is why lvref does not work. Also, we force the data_t type to be non-ref so there are no perfect-forwarding here. Pass-by-value is not good because that requires `data_t` to be copyable (but for there movable is enough).